presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -15,283 +15,367 @@
 
-            // scans
-            List<PlanNodeId> tableScans = fragment.getTableScanSchedulingOrder();
+        // TODO: ARBITRARY_DISTRIBUTION is something very weird.
+        // TODO: It doesn't have partitioning function, and it is never set as a fragment partitioning.
+        // TODO: We should consider removing ARBITRARY_DISTRIBUTION.
+        checkArgument(!partitioning.equals(ARBITRARY_DISTRIBUTION), "ARBITRARY_DISTRIBUTION is not expected to be set as a fragment distribution");
 
-            // source stages
-            List<RemoteSourceNode> remoteSources = fragment.getRemoteSourceNodes();
-            checkArgument(tableScans.isEmpty() || remoteSources.isEmpty(), "stages that have both, remote sources and table scans, are not supported");
+        int hashPartitionCount = getHashPartitionCount(session);
 
-            if (!tableScans.isEmpty()) {
-                checkArgument(fragment.getPartitioning().equals(SOURCE_DISTRIBUTION), "unexpected table scan partitioning: %s", fragment.getPartitioning());
+        // configure number of output partitions
+        if (fragment.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_HASH_DISTRIBUTION)) {
+            fragment = fragment.withBucketToPartition(Optional.of(IntStream.range(0, hashPartitionCount).toArray()));
+        }
 
-                // get all scheduled splits
-                List<ScheduledSplit> scheduledSplits = subPlan.getTaskSources().stream()
-                        .flatMap(taskSource -> taskSource.getSplits().stream())
-                        .collect(toImmutableList());
+        if (partitioning.equals(SINGLE_DISTRIBUTION) || partitioning.equals(FIXED_HASH_DISTRIBUTION) || partitioning.equals(FIXED_ARBITRARY_DISTRIBUTION)) {
+            checkArgument(
+                    fragment.getTableScanSchedulingOrder().isEmpty(),
+                    "Fragment with is not expected to have table scans. fragmentId: %s, fragment partitioning %s",
+                    fragment.getId(),
+                    fragment.getPartitioning());
+
+            for (RemoteSourceNode remoteSource : fragment.getRemoteSourceNodes()) {
+                if (remoteSource.isEnsureSourceOrdering() || remoteSource.getOrderingScheme().isPresent()) {
+                    throw new PrestoException(NOT_SUPPORTED, format(
+                            "Order sensitive exchange is not supported by Presto on Spark. fragmentId: %s, sourceFragmentIds: %s",
+                            fragment.getId(),
+                            remoteSource.getSourceFragmentIds()));
+                }
+            }
 
-                // get scheduled splits by task
-                List<List<ScheduledSplit>> assignedSplits = assignSplitsToTasks(scheduledSplits, initialSparkPartitionCount);
+            Partitioner inputPartitioner = createPartitioner(
+                    partitioning,
+                    // TODO: consider using getMaxTasksPerStage
```

Comment:
I am not sure. If so then we should also use `getMaxTasksPerStage` for the initial partition count : https://github.com/prestodb/presto/blob/472538a02722c5409649533f9a602b5b76d9dd2d/presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java#L161-L162

Reply:
I agree this comment is somehow confusing. Let me remove it for now.